File: 백준/Silver/12852. 1로 만들기 2/1로 만들기 2.cc
#include<bits/stdc++.h>
using namespace std;
int dp[1000004], n;
const int INF = 1e9;
void go(int n){
    if(n == 0) return;
    cout << n << ' ';
    if(n % 3 == 0 && dp[n] == (dp[n / 3] + 1))
        go(n / 3);
    else if(n % 2 == 0 && dp[n] == (dp[n / 2] + 1))
        go(n / 2);
    else if(n - 1 >= 0 && dp[n] == (dp[n - 1] + 1))
        go(n - 1);
    return;
}
int main(){
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    cin >> n;
    fill(dp, dp + 1000004, INF);
    dp[1] = 0;
    for(int i = 1; i <= n; i++){
        if(!(i % 3)) dp[i] = min(dp[i / 3] + 1, dp[i]);
        if(!(i % 2)) dp[i] = min(dp[i / 2] + 1, dp[i]);
        dp[i] = min(dp[i - 1] + 1, dp[i]);
    }
    cout << dp[n] << '\n';
    go(n);
    return 0;
}

File: 백준/Silver/12852. 1로 만들기 2/1로 만들기 2.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] dp = new int[1000004];
    static final int INF = 1000000000;

    static void go(int n) {
        if (n == 0) return;
        System.out.print(n + " ");

        if (n % 3 == 0 && dp[n] == dp[n / 3] + 1) {
            go(n / 3);
        } else if (n % 2 == 0 && dp[n] == dp[n / 2] + 1) {
            go(n / 2);
        } else if (n - 1 >= 0 && dp[n] == dp[n - 1] + 1) {
            go(n - 1);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        Arrays.fill(dp, INF);
        dp[1] = 0;

        for (int i = 1; i <= n; i++) {
            if (i % 3 == 0) dp[i] = Math.min(dp[i / 3] + 1, dp[i]);
            if (i % 2 == 0) dp[i] = Math.min(dp[i / 2] + 1, dp[i]);
            dp[i] = Math.min(dp[i - 1] + 1, dp[i]);
        }

        System.out.println(dp[n]);
        go(n);
    }
}